nautobot/core/api/views.py

```diff
@@ -608,6 +611,41 @@ class NautobotSpectacularRedocView(APIVersioningGetSchemaURLMixin, SpectacularRe
     """Extend SpectacularRedocView to support Nautobot's ?api_version=<version> query parameter."""
 
 
+@method_decorator(gzip_page, name="dispatch")
+class NautobotSpectacularAPIView(SpectacularAPIView):
+    def _get_schema_response(self, request):
+        # version specified as parameter to the view always takes precedence. after
+        # that we try to source version through the schema view's own versioning_class.
+        version = self.api_version or request.version or self._get_version_parameter(request)
+        cache_key = f"openapi_schema_cache_{version}_{settings.VERSION}"  # Invalidate cache on Nautobot release
+        etag = f'W/"{hash(cache_key)}"'
+
+        # With combined browser cache and backend cache, we have three options:
+        # - cache expired on browser, but Etag is the same (no changes in nautobot) -> 70-100ms response
+        # - cache expired on browser, Etag is different, cache present on backend -> 400-600ms response
+        # - cache expired on browser, Etag is different, no cache on backend -> 3-4s response
+
+        if_none_match = request.META.get("HTTP_IF_NONE_MATCH", "")
+        if if_none_match == etag:
+            return Response(status=304)
+
+        schema = cache.get(cache_key)
```

Comment:
To avoid introducing issues like #446, we'll need to make sure to invalidate this cache when creating/updating/deleting CustomField and Relationship records. Should be able to use some of the existing extras signals for this purpose.